### ISSUE TITLE
Firefox hardware video acceleration failes due to missing kcmp() system call

### ISSUE THREAD

**Issue by blitz** [NixOS/nixpkgs#112082]
**Describe the bug**

To enable hardware video decoding in Firefox the `kcmp()` system call needs to be available, because mesa depends on it. This was observed on an AMD Renoir Thinkpad.

To make this work, the `CHECKPOINT_RESTORE` config option needs to be enabled in Linux.

**To Reproduce**

1. On a system with AMD GPU with `firefox-wayland`, follow the [Arch Wiki](https://wiki.archlinux.org/index.php/Firefox#Hardware_video_acceleration) to enable the right `about:config` flags for VA-API for Firefox
2. Run firefox with debug logs: `MOZ_LOG="PlatformDecoderModule:5" firefox`
3. See that VA-API fails to be enabled by finding`VA-API FFmpeg is disabled by platform` in the log when you try to play a youtube video

**Expected behavior**
VA-API works and video decoding is hardware accelerated.

**Additional context**

The issue is linked to Mesa not being able to use the `kcmp()` system call. This in turn leads to VA-API not being initialzed. See the big `Note` at the beginning of https://wiki.archlinux.org/index.php/Firefox#Hardware_video_acceleration

The `kcmp()` system call is exposed by the Linux kernel when the `CHECKPOINT_RESTORE` config option is enabled. On NixOS it defaults to off. On a Fedora box, it defaults to on.

`kcmp()` usage seems to have been introduced in mesa in 2020 with this commit:

```
commit f76cbc7901f7d500f5a4f74aedfd29970d1efd00
Author: Michel Dänzer <michel@daenzer.net>
Date:   Mon Jan 6 18:24:52 2020 +0100

    util: Add os_same_file_description helper
    
    Acked-by: Pierre-Eric Pelloux-Prayer <pierre-eric.pelloux-prayer@amd.com>
    Part-of: <https://gitlab.freedesktop.org/mesa/mesa/merge_requests/3202>
```

**Notify maintainers**

Firefox: @edolstra
Mesa: @primeos @vcunat 
Linux kernel: ??

**Metadata**
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
 - firefox
 - firefox-wayland
# a list of nixos modules affected by the problem
module:
```

**Comment by primeos** [comment]
> To make this work, the CHECKPOINT_RESTORE config option needs to be enabled in Linux.

Sounds good to me, thanks for the detailed analysis!
We currently set `CHECKPOINT_RESTORE` only for CRIU (66080ae4d8b0ae5830f2671d5e33cec7649a64ac, cc @bendlas) but setting `EXPERT` as well shouldn't be required anymore. And on Debian Buster `CHECKPOINT_RESTORE=y` is also set by default:
```console
$ grep CHECKPOINT /boot/config-4.19.0-14-amd64
CONFIG_CHECKPOINT_RESTORE=y
```

From `init/Kconfig`:
```
config CHECKPOINT_RESTORE
        bool "Checkpoint/restore support"
        select PROC_CHILDREN
        default n
        help
          Enables additional kernel features in a sake of checkpoint/restore.
          In particular it adds auxiliary prctl codes to setup process text,
          data and heap segment sizes, and a few additional /proc filesystem
          entries.

          If unsure, say N here.
```

So in `pkgs/os-specific/linux/kernel/common-config.nix` we should be able to replace:
```
    criu = optionalAttrs (features.criu or false) ({
      EXPERT              = yes;
      CHECKPOINT_RESTORE  = yes;
    } // optionalAttrs (features.criu_revert_expert or true) {
      RFKILL_INPUT          = option yes;
      HID_PICOLCD_FB        = option yes;
      HID_PICOLCD_BACKLIGHT = option yes;
      HID_PICOLCD_LCD       = option yes;
      HID_PICOLCD_LEDS      = option yes;
      HID_PICOLCD_CIR       = option yes;
      DEBUG_MEMORY_INIT     = option yes;
    });
```
with just `CHECKPOINT_RESTORE  = yes;` and an explanation that this is required for CRIU and Mesa.

**Comment by vcunat** [comment]
Sounds OK at a quick look, but it would be nice to test it fixes the acceleration before merging that.  (even though it seems more about mesa using the syscall than about firefox)

**Comment by blitz** [comment]
@vcunat 

> Sounds OK at a quick look, but it would be nice to test it fixes the acceleration before merging that. (even though it seems more about mesa using the syscall than about firefox)

Good call. I retested everything with the same settings with and without `CHECKPOINT_RESTORE` and the only difference at the moment is that without it Mesa complains when starting Firefox:

```
amdgpu: os_same_file_description couldn't determine if two DRM fds reference the same file description.
If they do, bad things may happen!
```

But VA-API works in both cases despite the ominous warning. I could have _sworn_ it didn't until I changed the kernel settings. :confused:

**Comment by blitz** [comment]
@vcunat What do you think. Should we enable kcmp anyway now that mesa seems to depend on it? It seems like the right thing to do.

**Comment by ajs124** [comment]
Coincidentally, upstream seems to be changing `kcmp()` to be available without `CHECKPOINT_RESTORE`. See https://lwn.net/Articles/845448 (only freely available on February 25) or just look at the patchset: https://lwn.net/ml/linux-kernel/20210205220012.1983-1-chris@chris-wilson.co.uk/

This isn't to say that we shouldn't enable `CHECKPOINT_RESTORE`, just that it might not be necessary in the long run.

**Comment by vcunat** [comment]
I expect it's worth getting `kcmp()` by default.  I believe we'll be defaulting to 5.4 or 5.10 kernels for quite some time (about a year at least) and those branches surely won't change in this respect anymore.  (I assume they need `CHECKPOINT_RESTORE` for `kcmp()` but I haven't verified that.)

**Comment by ajs124** [comment]
We could just backport the patch ourselves, but since we don't really do that kind of thing AFAIK, we can just enable `CHECKPOINT_RESTORE` and maybe think about disabling it again, when an upstream stable kernel doesn't need it for `kcmp()` anymore.

Do we have any defined processes on how do decide on kernel config or is it as ad-hoc as everything else?

**Comment by vcunat** [comment]
I believe there's no process for that.

**Comment by blitz** [comment]
So `CHECKPOINT_RESTORE` was removed from `EXPERT` with this commit, which landed in Linux 4.19:

```
commit 5cb366bb3a746f6b06ea086b322e21e345401c9d
Author: Adrian Reber <adrian@lisas.de>
Date:   Tue Aug 21 22:01:17 2018 -0700

    init/Kconfig: remove EXPERT from CHECKPOINT_RESTORE
 ```

**Comment by blitz** [comment]
@vcunat @primeos I gave this a shot in #113236.
